### PR TITLE
Feature/add analysis tools

### PR DESCRIPTION
**Pull request by delyriand**

### CODE REVIEW

File: src/Repository/SettingRepository.php [hunk 1]
```diff
@@ -121,4 +78,48 @@ public function findAllByChannelAndLocaleWithDefault(string $vendor, string $plu
...
+        ;
+    }
+
+    /**
+     * @param mixed $value
+     */
+    private function addColumnConditionWithNull(QueryBuilder $queryBuilder, string $columnName, $value): void
+    {
+        $this->addColumnCondition($queryBuilder, $columnName, $value, sprintf(' OR o.%s IS NULL', $columnName));
+    }
```
Comment (maximehuran) [inline src/Repository/SettingRepository.php:124]: A lot of method for nothing no ?

File: src/Settings/Settings.php [hunk 2]
```diff
@@ -132,53 +106,38 @@ private function getCachedSettingsByChannelAndLocale(string $channelIdentifier,
...
     }
 
-    /**
-     * @param array $allSettings
-     *
-     * @return array
-     */
     private function stackSettings(array $allSettings): array
     {
+        dump($allSettings);
```
Comment (maximehuran) [inline src/Settings/Settings.php:138]: oopsy

File: src/Settings/Settings.php [hunk 3]
```diff
@@ -23,31 +23,12 @@ final class Settings implements SettingsInterface
...
-    /**
-     * @var array|null
-     */
-    private ?array $settingsByChannelAndLocale;
-
-    /**
-     * @var array|null
-     */
-    private ?array $settingsByChannelAndLocaleWithDefault;
-
```
Comment (maximehuran) [inline src/Settings/Settings.php:45]: Why it is removed ?

File: src/Settings/Settings.php [hunk 4]
```diff
@@ -95,32 +76,25 @@ public function getIcon(): string
...
-     * @return array|null
-     */
     private function getCachedSettingsByChannelAndLocale(string $channelIdentifier, string $localeIdentifier, bool $withDefault): ?array
     {
         // With default?
         $varName = $withDefault ? 'settingsByChannelAndLocaleWithDefault' : 'settingsByChannelAndLocale';
+        /** @phpstan-ignore-next-line */
         if (!isset($this->{$varName}[$channelIdentifier])) {
-            $this->{$varName}[$channelIdentifier] = [];
+            $this->{$varName}[$channelIdentifier] = []; /** @phpstan-ignore-line */
```
Comment (maximehuran) [inline src/Settings/Settings.php:97]: PHP Stan comment at the end of the line is what you want ?

File: src/Settings/Settings.php [hunk 5]
```diff
@@ -132,53 +106,38 @@ private function getCachedSettingsByChannelAndLocale(string $channelIdentifier,
...
-                        $this->metadata->getName(true),
-                        $channel,
-                        $localeCode
-                    )
-                );
-                $this->settingsByChannelAndLocale[$channelIdentifier][$localeIdentifier] = $settings;
-            }
+            $methodName = $withDefault ? 'findAllByChannelAndLocaleWithDefault' : 'findAllByChannelAndLocale';
+            $settings = $this->stackSettings(
+                $this->settingRepository->{$methodName}(
```
Comment (maximehuran) [inline src/Settings/Settings.php:120]: I am not fond of the using a dynamic method name

**Review by maximehuran — CHANGES_REQUESTED** [review]

File: src/Repository/SettingRepository.php (hunk 1 above)
Comment (delyriand) [reply src/Repository/SettingRepository.php]: yes, i wanted to avoid phpmd errors but the code is complicated

File: src/Settings/Settings.php (hunk 3 above)
Comment (delyriand) [reply src/Settings/Settings.php:45]: mistake ... with the refactoring of `getSettingsByChannelAndLocale` method, PHPStorm tells me that the variables are not used

File: src/Settings/Settings.php (hunk 4 above)
Comment (delyriand) [reply src/Settings/Settings.php]: with class properties, this instruction isn't necessary

File: src/Repository/SettingRepository.php [hunk 6]
```diff
@@ -121,4 +78,46 @@ public function findAllByChannelAndLocaleWithDefault(string $vendor, string $plu
...
+        }
+
+        $whereCondition = 'o.channel = :channel';
+        if ($withNull) {
+            $whereCondition .= ' OR o.channel IS NULL';
+        }
+        $queryBuilder
+            ->andWhere($whereCondition)
+            ->setParameter('channel', $channel)
+        ;
```
Comment (maximehuran) [inline src/Repository/SettingRepository.php:100]:
```suggestion
        if (null === $channel) {
            $queryBuilder->andWhere('o.channel IS NULL');
            return;
        }
        $queryBuilder
            ->andWhere('o.channel = :channel' . ($withNull ? ' OR o.channel IS NULL' : ''))
            ->setParameter('channel', $channel)
        ;
```

File: src/Repository/SettingRepository.php [hunk 7]
```diff
@@ -121,4 +78,46 @@ public function findAllByChannelAndLocaleWithDefault(string $vendor, string $plu
...
+        }
+
+        $whereCondition = 'o.localeCode = :localeCode';
+        if ($withNull) {
+            $whereCondition .= ' OR o.localeCode IS NULL';
+        }
+        $queryBuilder
+            ->andWhere($whereCondition)
+            ->setParameter('localeCode', $localeCode)
+        ;
```
Comment (maximehuran) [inline src/Repository/SettingRepository.php:121]:
```suggestion
        if (null === $localeCode) {
            $queryBuilder->andWhere('o.localeCode IS NULL');
            return;
        }
        $queryBuilder
            ->andWhere('o.localeCode = :localeCode' . ($withNull ? ' OR o.localeCode IS NULL' : ''))
            ->setParameter('localeCode', $localeCode)
        ;
```

**Review by maximehuran — CHANGES_REQUESTED** [review]
Just last changes

File: src/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -74,4 +66,14 @@ private function addPlugins(ArrayNodeDefinition $rootNode): void
         ->end()
         ;
     }
+
+    private function getRootNode(TreeBuilder $treeBuilder): ArrayNodeDefinition
+    {
+        if (method_exists($treeBuilder, 'getRootNode')) {
+            /** @phpstan-ignore-next-line */
+            return $treeBuilder->getRootNode();
+        }
```
Comment (jacquesbh) [inline src/DependencyInjection/Configuration.php:75]: Becomes useless since Sylius 1.8 required Symfony ^4.4.
Can be in another PR if you prefer.

File: src/Processor/SettingsProcessor.php [hunk 9]
```diff
@@ -136,22 +103,21 @@ private function saveSettings(SettingsInterface $settings, ?int $channelId, ?str
...
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
      */
     private function removeUnusedSettings(array &$data, array $settings): void
     {
         // Manage defaults, and remove actual settings with "use default value" checked
         foreach ($data as $key => $value) {
             // Is the setting a "use default value"?
             if (1 === preg_match(sprintf('`^(?P<key>.*)(?:___%1$s)$`', Settings::DEFAULT_KEY), $key, $matches)) {
-                if (true === $data[$key]) {
+                if (true === $value) {
```
Comment (jacquesbh) [inline src/Processor/SettingsProcessor.php:118]: Good catch 😅

**Review by jacquesbh — APPROVED** [review]

**Review by maximehuran — APPROVED** [review]

**Review by lanfisis — APPROVED** [review]

**Comment by lanfisis** [comment]
Thank you 😊
